### PR TITLE
CHE-1095: add new UD/IDE navigation experience

### PR DESCRIPTION
**Pull request by akurinnoy**
Goals:
- revamp the left navbar
- remove the mobile responsiveness
- add ability to jump into recent workspace
- improve the branding/consistency between UD/IDE

Signed-off-by: Oleksii Kurinnyi okurinnyi@codenvy.com

### CODE REVIEW

**Comment by akurinnoy** [comment]
navbar-che

File: core/ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/client/BootstrapController.java [hunk 1]
```diff
@@ -211,9 +211,22 @@ public void onClose(CloseEvent<Window> event) {
...
+        Scheduler.get().scheduleDeferred(new ScheduledCommand() {
+            @Override
+            public void execute() {
+                notifyIDELoaded();
+            }
+        });
     }
 
     /**
+     * Sends a message to the parent frame to inform that IDE ide has been loaded.
```
Comment (ashumilova) [inline core/ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/client/BootstrapController.java]: "IDE ide" -> "IDE application"

File: dashboard/src/app/dashboard/dashboard.controller.js [hunk 2]
```diff
@@ -23,9 +23,9 @@ export class DashboardController {
...
    */
-  constructor() {
+  constructor($rootScope) {
+    $rootScope.showIDE = false;
   }
-
-
 }
 
+DashboardController.$inject = ['$rootScope'];
```
Comment (benoitf) [inline dashboard/src/app/dashboard/dashboard.controller.js]: why do we need this inject ?

File: dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+   * Default constructor that is using resource
+   * @ngInject for Dependency injection
+   */
+  constructor($window) {
+    this.$window = $window;
+
+    this.offset = angular.isUndefined(this.offset) ? '0 0' : this.offset;
+  }
+
+  process(item) {
```
Comment (benoitf) [inline dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js:37]: doc about what is allowing process method and what is item

File: dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.directive.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ */
+'use strict';
+
+/**
+ * @author Oleksii Kurinnyi
+ */
+export class NavbarDropdownMenu {
```
Comment (benoitf) [inline dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.directive.js:19]: what is the goal of the class (comment)

File: dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ *   Codenvy, S.A. - initial API and implementation
+ */
+'use strict';
+
+
+/**
+ * This class is handling the controller for the dropdown menu on navbar
+ * @author Oleksii Kurinnyi
+ */
+export class NavbarDropdownMenuCtrl {
```
Comment (benoitf) [inline dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js]: controllers shouldn't match xxxxController classname now (to avoid lint errors)?

File: dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.directive.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        y = e.clientY,
+        stack = [];
+      let elementMouseIsOver = self.$document[0].elementFromPoint(x, y);
+      elementMouseIsOver.style.pointerEvents = 'none';
+      stack.push(elementMouseIsOver);
+
+      // iterate elements under cursor
+      let limit = 50,
+        nextTargetEl;
+      while (elementMouseIsOver.tagName !== 'MD-MENU' && limit>0){
```
Comment (benoitf) [inline dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.directive.js]: space formatting like "limit > 0"

File: dashboard/src/app/navbar/navbar.controller.js [hunk 7]
```diff
@@ -65,7 +66,7 @@ export class CheNavBarCtrl {
     $scope.$on('$locationChangeStart', () => {
       let path = '#' + $location.path(),
         match = Object.keys(this.menuItemUrl).some(item => this.menuItemUrl[item] === path);
-      if (match) {
+      if (match || /\/ide\//.test(path)) {
```
Comment (benoitf) [inline dashboard/src/app/navbar/navbar.controller.js]: is this pattern is working in both Che and SaaS ?

File: dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+ */
+'use strict';
+
+/**
+ * @ngdoc controller
+ * @name navbar.controller:NavbarRecentWorkspacesController
+ * @description This class is handling the controller of the recent workspaces to display in the navbar
+ * @author Oleksii Kurinnyi
+ */
+export class NavbarRecentWorkspacesCtrl {
```
Comment (benoitf) [inline dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js]: xxxxController pattern for the controller name

File: dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        _onclick: (workspaceId) => { this.runRecentWorkspace(workspaceId) }
+      }
+    ];
+    this.dropdownItems = {};
+
+    this.veryRecentWorkspaceId = '';
+    let cleanup = $rootScope.$on('recent-workspace:set', (event, workspaceId) => {
+      this.veryRecentWorkspaceId = workspaceId;
+    });
+    $rootScope.$on('$destroy', function() {
```
Comment (benoitf) [inline dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js]: "function()" may be replaced by "() =>" as in previous lines

File: dashboard/src/app/dashboard/dashboard.controller.js (hunk 2 above)
Comment (akurinnoy) [inline dashboard/src/app/dashboard/dashboard.controller.js]: DI with ngInject didn't work here

Comment (benoitf) [inline dashboard/src/app/dashboard/dashboard.controller.js]: if you use "ngInject"; in the constructor body ?

File: dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+  /**
+   * Returns name of workspace
+   * @returns {String}
+   */
+  getWorkspaceName(workspaceId) {
+    let workspace = this.cheWorkspace.getWorkspaceById(workspaceId);
+    return workspace ? workspace.config.name : 'unknown';
+  }
+
+  isOpen(workspaceId) {
```
Comment (benoitf) [inline dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js:116]: methods starting from here have no longer comments

**Comment by ashumilova** [comment]
License check fails - to fix execute:
`mvn license:format`

File: dashboard/src/assets/svg/dashboard.svg [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+        <g transform="translate(-11.000000, -67.000000)" fill="#808B9C">
+            <g id="Dashboard-Icon" transform="translate(11.000000, 67.000000)">
+                <rect id="Rectangle-17" x="0" y="8" width="6" height="6"></rect>
+                <rect id="Rectangle-17-Copy" x="8" y="8" width="6" height="6"></rect>
+                <rect id="Rectangle-17-Copy-3" x="0" y="0" width="6" height="6"></rect>
+                <rect id="Rectangle-17-Copy-2" x="8" y="0" width="6" height="6"></rect>
+            </g>
+        </g>
+    </g>
+</svg>
```
Comment (benoitf) [inline dashboard/src/assets/svg/dashboard.svg]: missing EOL

File: dashboard/src/assets/svg/factory.svg [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+            <g id="rocket" transform="translate(12.000000, 144.000000)">
+                <path d="M0.23853117,16 C0.23853117,16 0.00386450358,14.2753333 0.249197837,13.832 L1.9938645,10.6793333 C2.2518645,11.746 2.55919784,12.8606667 2.94119784,14.26 L0.23853117,16 Z M9.98719784,10.6426667 C9.72653117,11.7466667 9.4518645,12.7893333 9.07453117,14.27 L11.7625312,16 C11.7625312,16 11.9965312,14.276 11.7511978,13.832 L9.98719784,10.6426667 Z M5.0358645,16 L6.9658645,16 L7.3338645,14.6666667 L4.66719784,14.6666667 L5.0358645,16 L5.0358645,16 Z M6.00053117,0 C3.67453117,1.66866667 2.66719784,3.65866667 2.66719784,6.12733333 C2.66719784,8.354 3.43119784,10.9773333 4.2538645,14 L7.76919784,14 C8.55653117,10.888 9.31053117,8.282 9.3338645,6.104 C9.36053117,3.61666667 8.37319784,1.678 6.00053117,0 L6.00053117,0 Z M7.78319784,4.35933333 C7.41119784,4.562 6.00053117,4.66666667 6.00053117,4.66666667 C6.00053117,4.66666667 4.61919784,4.57 4.2398645,4.378 C4.53453117,3.37 5.10919784,2.49533333 6.00853117,1.69533333 C6.9278645,2.496 7.49919784,3.356 7.78319784,4.35933333 Z"></path>
+            </g>
+        </g>
+    </g>
+</svg>
```
Comment (benoitf) [inline dashboard/src/assets/svg/factory.svg]: missing EOL

File: dashboard/src/assets/svg/workspace.svg [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    <g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
+        <g transform="translate(-11.000000, -90.000000)" fill="#808B9C">
+            <g>
+                <g transform="translate(11.000000, 90.000000)">
+                    <path d="M0.00254545454,3.74875 L7,0 L13.9936364,3.7875 L7.00318182,7.45 L0.00254545454,3.74875 L0.00254545454,3.74875 Z M7.63636364,8.545625 L7.63636364,15 L14,11.64875 L14,5.195 L7.63636364,8.545625 L7.63636364,8.545625 Z M6.36363636,8.545625 L0,5.16375 L0,11.64875 L6.36363636,15 L6.36363636,8.545625 L6.36363636,8.545625 Z"></path>
+                </g>
+            </g>
+        </g>
+    </g>
+</svg>
```
Comment (benoitf) [inline dashboard/src/assets/svg/workspace.svg]: missing EOL

File: dashboard/src/components/api/che-workspace.factory.js [hunk 14]
```diff
@@ -398,4 +399,10 @@ export class CheWorkspace {
       });
     }
   }
+
+  createSnapshot(workspaceId) {
```
Comment (benoitf) [inline dashboard/src/components/api/che-workspace.factory.js:436]: doc + we may return without intermediate variable

File: plugins/plugin-dashboard/che-plugin-ext-dashboard/pom.xml [hunk 15]
```diff
@@ -26,16 +26,25 @@
...
+        </dependency>
+        <dependency>
             <groupId>com.google.inject</groupId>
             <artifactId>guice</artifactId>
         </dependency>
         <dependency>
             <groupId>org.eclipse.che.core</groupId>
-            <artifactId>che-core-ide-api</artifactId>
+            <artifactId>che-core-api-workspace-shared</artifactId>
+            <version>4.4.0-RC1-SNAPSHOT</version>
```
Comment (benoitf) [inline plugins/plugin-dashboard/che-plugin-ext-dashboard/pom.xml]: -1 here
version should not be set

File: dashboard/src/app/dashboard/dashboard.controller.js (hunk 2 above)
Comment (akurinnoy) [inline dashboard/src/app/dashboard/dashboard.controller.js]: it works, thanks

File: dashboard/src/app/ide/ide-iframe/ide-iframe.service.js [hunk 16]
```diff
@@ -22,32 +22,44 @@ class IdeIFrameSvc {
...
    * @ngInject for Dependency injection
    */
-  constructor ($timeout, $compile) {
+  constructor ($timeout, $compile, $rootScope, $location, $window) {
     this.iframeAdded = false;
     this.$timeout = $timeout;
     this.$compile = $compile;
 
+    $window.addEventListener("message", function(event) {
+      if ("ide-loaded" == event.data) {
```
Comment (ashumilova) [inline dashboard/src/app/ide/ide-iframe/ide-iframe.service.js]: dealt to use strict equals

File: dashboard/src/app/workspaces/workspace-status-indicator/workspace-status-indicator-circle.directive.js [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ */
+'use strict';
+
+/**
+ * Defines a directive for creating simple circle indicator of workspace's status.
+ * @author Oleksii Kurinnyi
+ */
+export class WorkspaceStatusIndicatorCircle {
```
Comment (ashumilova) [inline dashboard/src/app/workspaces/workspace-status-indicator/workspace-status-indicator-circle.directive.js]: WorkspaceStatusIndicatorCircle -> WorkspaceStatusIndicator cause can be not only circle (in progress example)

File: dashboard/src/app/workspaces/workspace-status-indicator/workspace-status-indicator-circle.directive.js [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+    this.replace = true;
+
+    this.scope = {
+      status: '=cheStatus'
+    };
+  }
+
+  /**
+   * Template for the current toolbar
```
Comment (ashumilova) [inline dashboard/src/app/workspaces/workspace-status-indicator/workspace-status-indicator-circle.directive.js]: "current toolbar" - seems copy/paste error

File: dashboard/src/assets/svg/project.svg [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+-->
+
+<svg
+   xmlns:dc="http://purl.org/dc/elements/1.1/"
+   xmlns:cc="http://creativecommons.org/ns#"
+   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
+   xmlns:svg="http://www.w3.org/2000/svg"
+   xmlns="http://www.w3.org/2000/svg"
+   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
+   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
```
Comment (ashumilova) [inline dashboard/src/assets/svg/project.svg]: no need to keep inkscape data, same for following part:
`<metadata id="metadata20"><rdf:RDF>....`

File: dashboard/src/components/attribute/click/che-on-right-click.directive.js [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+   * Keep reference to the model controller
+   */
+  link($scope, $element, attrs) {
+    $element.bind('contextmenu', (event) => {
+      event.stopPropagation();
+      event.preventDefault();
+
+      $scope.$apply(() => {
+        $scope.$eval(attrs.cheOnRightClick);
+        $scope.longTouch = false;
```
Comment (ashumilova) [inline dashboard/src/components/attribute/click/che-on-right-click.directive.js]: $scope.longTouch = false; - seems to be part of other directive - `CheOnLongTouch`

File: dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js (hunk 5 above)
Comment (olexii4) [inline dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js]: +1 But it should be a separate commit(maybe, separate PR) to all of inclusions like  xxxxCtrl.

File: dashboard/src/app/navbar/navbar.controller.js (hunk 7 above)
Comment (akurinnoy) [inline dashboard/src/app/navbar/navbar.controller.js]: yes, it is
but I can slightly simplify it and remove this condition at all.

File: core/ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/client/BootstrapController.java (hunk 1 above)
Comment (akurinnoy) [inline core/ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/client/BootstrapController.java]: fixed

File: dashboard/src/app/dashboard/dashboard.controller.js (hunk 2 above)
Comment (akurinnoy) [inline dashboard/src/app/dashboard/dashboard.controller.js]: fixed

File: dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js (hunk 3 above)
Comment (akurinnoy) [inline dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js:37]: fixed

**Comment by benoitf** [comment]
PR needs to be rebased (conflicts)

File: dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.directive.js (hunk 4 above)
Comment (akurinnoy) [inline dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.directive.js:19]: fixed

File: dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js (hunk 5 above)
Comment (akurinnoy) [inline dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.controller.js]: agree with @benoitf 
new controller should be named properly
fixed

File: dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.directive.js (hunk 6 above)
Comment (akurinnoy) [inline dashboard/src/app/navbar/navbar-dropdown-menu/navbar-dropdown-menu.directive.js]: fixed

File: dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js (hunk 8 above)
Comment (akurinnoy) [inline dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js]: fixed

File: dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js (hunk 9 above)
Comment (akurinnoy) [inline dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js]: fixed

File: dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js (hunk 10 above)
Comment (akurinnoy) [inline dashboard/src/app/navbar/recent-workspaces/recent-workspaces.controller.js:116]: fixed

File: dashboard/src/assets/svg/dashboard.svg (hunk 11 above)
Comment (akurinnoy) [inline dashboard/src/assets/svg/dashboard.svg]: fixed

File: dashboard/src/assets/svg/factory.svg (hunk 12 above)
Comment (akurinnoy) [inline dashboard/src/assets/svg/factory.svg]: fixed

File: dashboard/src/assets/svg/workspace.svg (hunk 13 above)
Comment (akurinnoy) [inline dashboard/src/assets/svg/workspace.svg]: fixed

File: dashboard/src/components/api/che-workspace.factory.js (hunk 14 above)
Comment (akurinnoy) [inline dashboard/src/components/api/che-workspace.factory.js:436]: fixed

File: plugins/plugin-dashboard/che-plugin-ext-dashboard/pom.xml (hunk 15 above)
Comment (akurinnoy) [inline plugins/plugin-dashboard/che-plugin-ext-dashboard/pom.xml]: fixed

File: dashboard/src/app/colors/che-colors.styl [hunk 21]
```diff
@@ -55,6 +55,19 @@ $che-loader-active-title-color = $che-navy-color
...
+$navbar-border-right-color = #000000
+$navbar-active-menu-mark-color = $che-logo-yellow-color
+$navbar-menu-icon-color = #808B9C
+$navbar-active-menu-icon-color = $light-gray-color
+$navbar-menu-text-color = $mouse-gray-color
+$navbar-menu-dark-text-color = $navbar-menu-icon-color
+$navbar-active-menu-text-color = $light-gray-color
+$navbar-section-title-color = #62687C
+$navbar-ide-iframe-button-background-color = lighten($navbar-backgroud-color, 31%)
+$navbar-ide-iframe-button-border-color = alpha($navbar-ide-iframe-button-background-color, 0.65)
```
Comment (olexii4) [inline dashboard/src/app/colors/che-colors.styl]: Why these constants hardcoded there (not in che-colors.json)?

File: dashboard/src/app/ide/ide-iframe/ide-iframe.service.js (hunk 16 above)
Comment (akurinnoy) [inline dashboard/src/app/ide/ide-iframe/ide-iframe.service.js]: fixed

File: dashboard/src/app/colors/che-colors.styl [hunk 22]
```diff
@@ -55,6 +55,19 @@ $che-loader-active-title-color = $che-navy-color
...
+$navbar-backgroud-color = #282E44
+$navbar-active-menu-background-color = alpha($white-color, 20%)
+$navbar-border-right-color = #000000
+$navbar-active-menu-mark-color = $che-logo-yellow-color
+$navbar-menu-icon-color = #808B9C
+$navbar-active-menu-icon-color = $light-gray-color
+$navbar-menu-text-color = $mouse-gray-color
+$navbar-menu-dark-text-color = $navbar-menu-icon-color
+$navbar-active-menu-text-color = $light-gray-color
+$navbar-section-title-color = #62687C
```
Comment (olexii4) [inline dashboard/src/app/colors/che-colors.styl]: Why these constants hardcoded there (not in che-colors.json)?

File: dashboard/src/app/workspaces/workspace-status-indicator/workspace-status-indicator-circle.directive.js (hunk 17 above)
Comment (akurinnoy) [inline dashboard/src/app/workspaces/workspace-status-indicator/workspace-status-indicator-circle.directive.js]: fixed

File: dashboard/src/app/workspaces/workspace-status-indicator/workspace-status-indicator-circle.directive.js (hunk 18 above)
Comment (akurinnoy) [inline dashboard/src/app/workspaces/workspace-status-indicator/workspace-status-indicator-circle.directive.js]: fixed

File: dashboard/src/assets/svg/project.svg (hunk 19 above)
Comment (akurinnoy) [inline dashboard/src/assets/svg/project.svg]: fixed

File: dashboard/src/components/attribute/click/che-on-right-click.directive.js (hunk 20 above)
Comment (akurinnoy) [inline dashboard/src/components/attribute/click/che-on-right-click.directive.js]: fixed

File: dashboard/src/app/colors/che-colors.styl (hunk 22 above)
Comment (akurinnoy) [inline dashboard/src/app/colors/che-colors.styl]: I have added black color into che-colors.json .
Other colors are belong to specific theme of navbar and it doesn't make sense to put them to general colors list. So I move them into separate .styl file.
